### ISSUE TITLE
Avoid scanning all references for DynamicInterfaceCastableImplementation

### ISSUE THREAD

**Issue by sbomer** [mono/linker#1788]
We currently scan all references for interfaces with `DynamicInterfaceCastableImplementationAttribute`, and keep any that implement marked interfaces. This attribute was designed to make the `IDynamicCastableInterface` feature linker-friendly, by explicitly annotating types that can serve as interface implementations for dynamic casts.

Depending on the expected behavior of `GetInterfaceImplementation`, we may be able to avoid this scan. There are already cases where we would not mark the attributed implementation (when `GetInterfaceImplementation` does non-analyzable reflection to get the implementing type's assembly, and that assembly isn't statically referenced).

However, we currently still scan *referenced* assemblies and keep such types (even if they are only used by non-analyzable reflection). Would it be safe to remove the attributed interfaces in this case? The linker already issues a warning for this.

**Comment by MichalStrehovsky** [comment]
Yes, I think that was the gist of how I was thinking to implement it in #1244 - "If an interface with DynamicInterfaceCastableImplementationAttribute is marked, do the same thing as if it was implemented by an instantiated type."

I didn't look at how the handling was implemented, but if we keep all interfaces marked as DynamicInterfaceCastableImplementationAttribute just because they require a marked interface, that seems more than what's strictly needed.

We should be able to postpone checking for DynamicInterfaceCastableImplementationAttribute until the interface is marked.
